File: kotlin/src/main/kotlin/382.LinkedListRandomNode.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/linked-list-random-node/


class LinkedListRandomNode{
  class Solution(val head: ListNode?) {

    /** @param head The linked list's head.
    Note that the head is guaranteed to be not null, so it contains at least one node. */

    val rand = Random()

    /** Returns a random node's value. */
    fun getRandom(): Int {
      if(head == null) return -1
      var cur = head!!
      var chosen = cur
      var len = 1
      while(cur.next != null){
        cur = cur.next!!
        len += 1
        if(rand.nextInt(len) == 0)
          chosen = cur
      }
      return chosen.`val`
    }

  }
}

File: python/382_LinkedListRandomNode.py
from unittest import TestCase
# https://leetcode.com/problems/linked-list-random-node/
from Definitions import ListNode
import random


class Solution:

    def __init__(self, head):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        :type head: ListNode
        """
        self.head = head

    def getRandom(self):
        """
        Returns a random node's value.
        :rtype: int
        """
        cur = self.head
        chosen = None
        size = 0
        while cur:
            size += 1
            if random.randint(1, size) == 1:
                chosen = cur
            cur = cur.next
        return chosen.val


